dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java

```diff
@@ -64,51 +125,737 @@ The former is disallowed by assumption, so we have co(w3, w2) and hence c(w3) <
+                must.add(new Tuple(e, e.getSuccessor().getSuccessor()));
+            }
+            // Atomics blocks: BeginAtomic -> EndAtomic
+            for (Event end : program.getCache().getEvents(
+                    FilterIntersection.get(FilterBasic.get(RMW), FilterBasic.get(SVCOMP.SVCOMPATOMIC)))) {
+                List<Event> block = ((EndAtomic) end).getBlock().stream().filter(x -> x.is(VISIBLE)).collect(toList());
+                for (int i = 0; i < block.size(); i++) {
+                    Event e = block.get(i);
+                    for (int j = i + 1; j < block.size(); j++) {
+                        if (!exec.areMutuallyExclusive(e, block.get(j))) {
+                            must.add(new Tuple(e, block.get(j)));
+                        }
+                    }
+                }
+            }
+            // ----- Compute maxTupleSet -----
+            TupleSet may = new TupleSet(must);
+            // LoadExcl -> StoreExcl
+            for (Thread thread : program.getThreads()) {
+                List<Event> events = thread.getCache().getEvents(FilterBasic.get(EXCL));
+                // assume order by cId
+                // assume cId describes a topological sorting over the control flow
+                for (int end = 1; end < events.size(); end++) {
+                    Event store = events.get(end);
+                    if (!store.is(WRITE)) {
+                        continue;
+                    }
+                    int start = iterate(end - 1, i -> i >= 0, i -> i - 1)
+                            .filter(i -> exec.isImplied(store, events.get(i)))
+                            .findFirst().orElse(0);
+                    List<Event> candidates = events.subList(start, end).stream()
+                            .filter(e -> !exec.areMutuallyExclusive(e, store))
+                            .collect(toList());
+                    int size = candidates.size();
+                    for (int i = 0; i < size; i++) {
+                        Event load = candidates.get(i);
+                        List<Event> intermediaries = candidates.subList(i + 1, size);
+                        if (!load.is(READ) || intermediaries.stream().anyMatch(e -> exec.isImplied(load, e))) {
+                            continue;
+                        }
+                        Tuple tuple = new Tuple(load, store);
+                        may.add(tuple);
+                        if (intermediaries.stream().allMatch(e -> exec.areMutuallyExclusive(load, e)) &&
+                                (store.is(MATCHADDRESS) || alias.mustAlias((MemEvent) load, (MemEvent) store))) {
+                            must.add(tuple);
+                        }
+                    }
+                }
+            }
+            return new Knowledge(may, must);
+        }
+        @Override
+        public Knowledge visitMemoryOrder(Relation rel) {
+            logger.info("Computing maxTupleSet for memory order");
```

Comment:
I think logging shouldn't really be done inside those methods at all.
It would be fine to have the Relation Analysis finish, and then iterate through all interesting relations (rf, co, rmw and possibly the axiom relations) and output their sizes in a central place.
Logging within those methods may make sense if they compute many intermediate results and we want to observe those, but then I would likely do this with trace-level logging.

Reply:
I'd like to move those changes into #330, to avoid conflicts.  The logging time stamps were useful in debugging this and the other PR.

Reply:
#330 now includes these changes.

Reply:
Now this PR also has them.